### PR TITLE
Add "override" keyword for overridden virtual functions

### PR DESCRIPTION
**Pull request by khuzemap**
Add PROTOBUF_OVERRIDE macro, which is defined to be
"override" for c++11 and nothing for older versions.
PROTOBUF_OVERRIDE is added for overridden virtual
functions.
Please refer following issue for discussion on this.
https://github.com/google/protobuf/issues/67

### CODE REVIEW

**Comment by xfxyjwf** [comment]
@khuzemap I just commented on the issue: we are allowing C++11 features in protobuf master now. Could you update this PR accordingly?

File: src/google/protobuf/compiler/cpp/cpp_message.cc [hunk 1]
```diff
@@ -1121,12 +1122,12 @@ GenerateClassDefinition(io::Printer* printer) {
   if (HasGeneratedMethods(descriptor_->file(), options_)) {
     if (HasDescriptorMethods(descriptor_->file(), options_)) {
       printer->Print(vars,
-        "void CopyFrom(const ::google::protobuf::Message& from) PROTOBUF_FINAL;\n"
-        "void MergeFrom(const ::google::protobuf::Message& from) PROTOBUF_FINAL;\n");
+        "void CopyFrom(const ::google::protobuf::Message& from) PROTOBUF_FINAL PROTOBUF_OVERRIDE;\n"
```
Comment (xfxyjwf) [inline src/google/protobuf/compiler/cpp/cpp_message.cc]: From what I understand you don't need "override" if you already have "final". Doesn't "final" imply "override" already?

Comment (khuzemap) [reply src/google/protobuf/compiler/cpp/cpp_message.cc]: No. "override" and "final" have independent behavior. "final" means the function cannot be overridden, however, "override" means the function is overriding a base class function. If a function has both "override" and "final" it means, the function is overriding a base class function but cannot be overriden.

Comment (xfxyjwf) [reply src/google/protobuf/compiler/cpp/cpp_message.cc]: I believe "final" has the exactly same semantics as "final override", i.e., it's overriding a virtual method in a base class and it can no longer be overridden.  In any case, please follow our C++ style guide:
https://google.github.io/styleguide/cppguide.html#Inheritance

Specifically:
```
Explicitly annotate overrides of virtual functions or virtual destructors with exactly one of an override or (less frequently) final specifier
```

Comment (khuzemap) [reply src/google/protobuf/compiler/cpp/cpp_message.cc]: we use gcc "Wsuggest-override" warning option for compile time enforcement and that doesn't take "final" as replacement of "override". clang has better option called "Winconsistent-missing-override", but unfortunately not available on gcc yet.

Comment (xfxyjwf) [reply src/google/protobuf/compiler/cpp/cpp_message.cc]: I think we should follow the style guide here. I just checked Google's internal code base and there is pretty much 0 usage of "final override" and "override final". Using it in protobuf code would require an exception granted by the style guide arbiters but I don't think it's justified for this case. Eventually gcc will have the same option, won't it?

**Comment by xfxyjwf** [comment]
@khuzemap Do you plan to update the PR to conform to Google's style guide? (i.e., use only one of "final" and "override").

**Comment by khuzemap** [comment]
@xfxyjwf , sorry for not pushing it early. have updated the change with only one of override or final. please review. Thanks!

File: src/google/protobuf/io/zero_copy_stream_impl.h [hunk 2]
```diff
@@ -67,6 +67,7 @@ class LIBPROTOBUF_EXPORT FileInputStream : public ZeroCopyInputStream {
   // should be read and returned with each call to Next().  Otherwise,
   // a reasonable default is used.
   explicit FileInputStream(int file_descriptor, int block_size = -1);
+  ~FileInputStream() override = default;
```
Comment (xfxyjwf) [inline src/google/protobuf/io/zero_copy_stream_impl.h]: Why is this needed?

File: src/google/protobuf/map_field.h [hunk 3]
```diff
@@ -311,27 +312,28 @@ class LIBPROTOBUF_EXPORT DynamicMapField: public TypeDefinedMapFieldBase<MapKey,
...
  private:
   Map<MapKey, MapValueRef> map_;
   const Message* default_entry_;
 
   // Implements MapFieldBase
-  void SyncRepeatedFieldWithMapNoLock() const;
-  void SyncMapWithRepeatedFieldNoLock() const;
-  size_t SpaceUsedExcludingSelfNoLock() const;
-  void SetMapIteratorValue(MapIterator* map_iter) const;
+
```
Comment (xfxyjwf) [inline src/google/protobuf/map_field.h]: Remove this empty line

File: src/google/protobuf/compiler/importer.h [hunk 4]
```diff
@@ -119,13 +120,13 @@ class LIBPROTOBUF_EXPORT SourceTreeDescriptorDatabase : public DescriptorDatabas
...
                   ErrorLocation location,
-                  const string& message);
+                  const string& message) override;
 
     virtual void AddWarning(const string& filename,
-                            const string& element_name,
-                            const Message* descriptor,
-                            ErrorLocation location,
-                            const string& message);
+                    const string& element_name,
```
Comment (xfxyjwf) [inline src/google/protobuf/compiler/importer.h]: Revert this alignment change.

Comment (khuzemap) [reply src/google/protobuf/compiler/importer.h]: Done

Comment (khuzemap) [reply src/google/protobuf/compiler/importer.h]: Done

File: src/google/protobuf/map_field.h (hunk 3 above)
Comment (khuzemap) [reply src/google/protobuf/map_field.h]: Done

File: src/google/protobuf/io/zero_copy_stream_impl.h (hunk 2 above)
Comment (khuzemap) [reply src/google/protobuf/io/zero_copy_stream_impl.h]: removed.

**Comment by xfxyjwf** [comment]
The MacOS C++ test is failing:
```
./google/protobuf/map_entry_lite.h:275:10: error: 'SpaceUsedLong' overrides a member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]
  size_t SpaceUsedLong() const {
```

Can you fix that?

**Comment by khuzemap** [comment]
MapEntryImpl is templated and instantiated with its Base class as Message and MessageLite. Message class defines SpaceUsedLong as virtual but MessageLite class don't. So enforcing 'override' here fails when MapEntryImpl is instantiated with MessageLite. Since there was enforcement of 'override', I didn't add any special handling. If "override" is enforced on code base, then for this function, it needs to be waived/ignored. In our code we use "-Wsuggest-override" to enforce "override" so I have added code to ignore that. For macOS build, it seems -Winconsistent-missing-override is used, so it needs to be ignored. of course other option is not to use any compile time switch as there are more than one ways of doing it.

**Comment by xfxyjwf** [comment]
Can we move that SpaceUsedLong() method from MapEntryImpl to MapEntry?

**Comment by khuzemap** [comment]
Was able to move. here is the diff.

**Comment by xfxyjwf** [comment]
@khuzemap Can you update this pull request?

**Comment by khuzemap** [comment]
yes, done.

**Comment by xfxyjwf** [comment]
I'm rerunning the tests.

**Comment by daquexian** [comment]
This "Down-integrate internal changes to github" commit https://github.com/protocolbuffers/protobuf/commit/624a40a387e1b2da4c6bbde44ad3d0a5efb9879b reverted this change :(
